File: src/benchmarks/test_cases/security/sql_injection.py
"""包含 SQL 注入漏洞的示例代码"""
import sqlite3


def get_user_by_name(username):
    """危险：直接拼接 SQL，存在 SQL 注入漏洞"""
    conn = sqlite3.connect("users.db")
    cursor = conn.cursor()
    # SQL 注入：username 未经过滤直接拼入查询
    query = f"SELECT * FROM users WHERE name = '{username}'"
    cursor.execute(query)
    return cursor.fetchall()


def get_user_by_id(user_id):
    """危险：格式化字符串构造 SQL"""
    conn = sqlite3.connect("users.db")
    # SQL 注入
    query = "SELECT * FROM users WHERE id = %s" % user_id
    return conn.execute(query).fetchall()


# 硬编码密钥
API_KEY = "benchmark_fake_api_key"
DATABASE_PASSWORD = "admin123456"


def execute_command(cmd):
    """危险：命令注入"""
    import os
    os.system(cmd)

File: src/benchmarks/test_cases/java/sql_injection.java
/** 包含 SQL 注入和安全问题的 Java 示例 */
import java.sql.*;

public class UserService {
    private static final String DB_PASSWORD = "admin123456";
    private static final String API_KEY = "benchmark_fake_api_key";

    public User getUser(String username) throws SQLException {
        Connection conn = DriverManager.getConnection("jdbc:mysql://localhost/db", "root", DB_PASSWORD);
        // SQL 注入：字符串拼接
        String query = "SELECT * FROM users WHERE name = '" + username + "'";
        Statement stmt = conn.createStatement();
        ResultSet rs = stmt.executeQuery(query);
        return mapUser(rs);
    }

    public User getUserById(int id) throws SQLException {
        Connection conn = DriverManager.getConnection("jdbc:mysql://localhost/db", "root", "pass");
        // SQL 注入：String.format 构造查询
        String query = String.format("SELECT * FROM users WHERE id = %d", id);
        return conn.createStatement().executeQuery(query);  // 未关闭连接
    }

    // 硬编码密码
    public boolean login(String email, String password) {
        return password.equals("masterkey123");
    }
}
